### PR TITLE
ADT-449: Merge Github and Github actions extensions

### PR DESCRIPTION
**Pull request by jemmyw**
This PR merges in the GitHub Actions extension from https://github.com/aha-develop/github-actions so that Actions is just an additional attribute.

I found the existing GitHub code hard to follow while adding to it. The problem was that we were using the different shapes of a pull request interchangeably. I found at least 4 different kinds of "pull request": the PR data we store in the extension fields, the PR webhook event, other webhooks that referred to PRs and the PR data from the github graphql API. Further, the graphql API could be given options to include more information. In places like the react component that renders the PR for the attribute we pass one type then later replace it with another type, and as the fields are different this led to lots of hard to understand conditionals.

### Converters

I have changed the code so that each component that renders something renders from only one source of data. The file `lib/github/converters.ts` provides functions that turn each remote source of github data into the shape we store in the extension, for example:

https://github.com/aha-develop/github/blob/39cc915e5cb1dd0f036591249d748f745141cbfc/src/lib/github/converters.ts#L27-L38

I could then change all the code that managed PRs to fetch -> convert -> link and the types provided by the octokit schema ensure we're always passing the right thing around.

### Codegen

Previously we were maintaining a lot of type information about the github shape data and graphql queries. But github have all this type information available in the octokit packages. Using the codegen package https://the-guild.dev/graphql/codegen we can instead generate the exact types we use from the schema with our queries. The queries are now in `.graphql` files in src/lib/github/queries

codegen generates an parsed version of these queries into src/generated/graphql.ts. This is a large file, but most of it is type data so it does not get bundled. By implementing the GqlFetch interface as a wrapper for the octokit graphql function here https://github.com/aha-develop/github/blob/4648c6273c891e52e1487e154e7fe62b2007f66e/src/lib/github/api.ts#L11 we can call the API with a precompiled query and get types for the expected variables and the return data.

So at the expense of some extra setup, we get better types that we don't have to maintain ourselves.

### Actions extension field data

Moving the actions attribute into this extension means the extenison data is lost. I considered if we should write a migration, but I'm of the opinion now that we shouldn't bother. I'm not 100% sure the data it was storing previously was correct as I had to fix up the code a bit, and it's not working anyway for customers due to the tooltip component changing.

This will be mitigated somewhat as the actions data will now also be fetched from github when the PR is linked or refreshed by the user.

### Pull request linking by commit message

We've never been able to link pull requests up by commit message because the PR events do not include the commit message. Nor do the branch create events that we've been using to link by branch. 

BUT the push event does contain the commit message, so I've come up with a way to use that. This code now listens for the `push` event and then it does the branch linking using the message. I've added a branch URL link to the account extension fields just like the one we use for pull requests: https://github.com/aha-develop/github/blob/f7b11fe3b03608ccfd9edabebc74c694f727fbb6/src/lib/linkBranch.ts#L39

Now when a PR is opened it can figure out the branch URL and perform a lookup: https://github.com/aha-develop/github/blob/f7b11fe3b03608ccfd9edabebc74c694f727fbb6/src/lib/recordFrom.ts#L153-L156

### Record from file

To make this linking a bit "easier" I've added a new file called `recordFrom` and moved all of the code into there that finds records using various other things. I've then made each of these log where it found the record. This way we can look at the invocation logs and see that a PR was found by title or url or whatever. This should be helpful for support.

### CODE REVIEW

**Comment by trydionel** [comment]
Great write-up @jemmyw! I'll give this a spin later today.

**Review by trydionel — COMMENTED** [review]
Initial happy-path run-through looks good. What else do you recommend I test?

- [ ] Ref # in PR title
- [ ] Ref # in branch
- [ ] Ref # in commit? (did we carry that over?)

File: src/components/Actions/Attribute.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      });
+
+  return (
+    <aha-flex
+      align-items="left"
+      direction="column"
+      gap="5px"
+      style={{ padding: "0 5px" }}
+    >
+      {hasWorkflows ? <>{workflows}</> : <EmptyState record={record} />}
```
Comment (trydionel) [inline src/components/Actions/Attribute.tsx:57]: Can we take this opportunity to use the improved empty state design? As defined in `src/components/attribute/EmptyState.tsx`

**Review by trydionel — CHANGES_REQUESTED** [review]
I encountered some issues while testing this:

* When putting the reference number in the branch name, the GitHub field showed the associated branch. When I opened a PR for that branch _without including the ref # in the PR name_, the GitHub field didn't update to include the PR. The GitHub Actions field showed the related build though. We used to check for the branch name to link PRs, right?
* When putting the reference number in the PR name, neither of the fields showed information from GitHub: both the GitHub and GitHub Actions fields were empty. Logs showed "No record found for this PR"
* When putting the reference number in the commit message, only the GitHub Actions field showed information from GitHub.

Can we make these experiences more consistent?

**Comment by jemmyw** [comment]
> I encountered some issues while testing this:
> 
> * When putting the reference number in the branch name, the GitHub field showed the associated branch. When I opened a PR for that branch _without including the ref # in the PR name_, the GitHub field didn't update to include the PR. The GitHub Actions field showed the related build though. We used to check for the branch name to link PRs, right?
> * When putting the reference number in the PR name, neither of the fields showed information from GitHub: both the GitHub and GitHub Actions fields were empty. Logs showed "No record found for this PR"
> * When putting the reference number in the commit message, only the GitHub Actions field showed information from GitHub.
> 
> Can we make these experiences more consistent? Ideally putting the reference number if any of the branch name, PR name, or commit message would provide a complete experience.

@trydionel Unfortunately we cannot do this. I’ve looked into it and there’s a good reason we don’t link to a PR by commit message. we don’t have any way to get that information from the webhooks. In the events we get the branch name, but not the message on it and we cannot retrieve it because the server side webhook handler doesn’t have access to the graphql API

**Comment by jemmyw** [comment]
> @trydionel Unfortunately we cannot do this. I’ve looked into it and there’s a good reason we don’t link to a PR by commit message. we don’t have any way to get that information from the webhooks. In the events we get the branch name, but not the message on it and we cannot retrieve it because the server side webhook handler doesn’t have access to the graphql API

I've figured something out for this, see the section "Pull request linking by commit message" in the PR description

File: README.md [hunk 2]
```diff
@@ -58,6 +58,12 @@ Install required modules:
 yarn install
 ```
 
+The extension makes requests to GitHub's graphql API. Changes to the graphql queries in lib/github/queries need to be compiled by running:
```
Comment (trydionel) [inline README.md:61]: We have a list of webhook events the extension needs listed above. Is this still accurate? 

> Select the following individual events: Branch or tag creation, Check runs, Pull requests, Pull request reviews, Pushes, Statuses. Enable the webhook.

**Review by trydionel — APPROVED** [review]
Works great now! Thank you @jemmyw. One question about whether the README is still accurate, but good to go otherwise.

File: README.md (hunk 2 above)
Comment (jemmyw) [reply README.md:61]: Yes it is, funnily enough we're already listing the events for actions and pushes that this change now uses but we didn't before. 🤷🏻

**Comment by jemmyw** [comment]
> Works great now! Thank you @jemmyw. One question about whether the README is still accurate, but good to go otherwise.

Yep README is coincidentally already accurate

**Review by percyhanna — COMMENTED** [review]
I just noticed a couple small things. I have also requested a review from @cornu-ammonis.

File: src/components/Actions/Attribute.tsx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    [actionFields]
+  );
+
+  const useableFields = Object.values(actionFields);
+  const workflows =
+    hasWorkflows &&
+    useableFields
+      .filter((field) => field.project)
+      .map((field, index) => {
+        if (!field.project) return;
```
[the remark this answers is not in the document]
Comment (jemmyw) [reply src/components/Actions/Attribute.tsx:37]: sure. not sure it makes a difference, undefined vs null?

[the remark this answers is not in the document]
Comment (jemmyw) [reply src/components/Actions/Attribute.tsx]: done. although I think this branch won't actually ever run

File: src/components/Actions/AttributeWorkflows.tsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      workflows.sort(
+        (a, b) =>
+          new Date(b.finishTime || 0).getTime() -
+          new Date(a.finishTime || 0).getTime()
+      ),
+    [workflows]
+  );
+
+  const workflowElements = sortedWorkflows.map((workflow, index) => (
+    <AttributeWorkflow workflow={workflow} index={index} />
```
[the remark this answers is not in the document]
Comment (jemmyw) [reply src/components/Actions/AttributeWorkflows.tsx:22]: yep, added

File: src/components/Actions/CardLabel.tsx [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+
+export type CardLabelProps = {
+  title: string;
+  value?: React.ReactNode;
```
[the remark this answers is not in the document]
Comment (jemmyw) [reply src/components/Actions/CardLabel.tsx:5]: For the title or the value? Seems like they have equal weighting so I'm thinking maybe we shouldn't use children?

File: src/lib/github/getStatusCommit.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import { PrCommitFragment } from "generated/graphql";
+
+/**
+ * Extract the last commit from a pull request as fetched from graphql
+ */
+export function getLastCommit(pr: PrCommitFragment) {
```
[the remark this answers is not in the document]
Comment (jemmyw) [reply src/lib/github/getStatusCommit.ts]: done

File: src/components/Actions/CardLabel.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (percyhanna) [reply src/components/Actions/CardLabel.tsx:5]: Ya, fair point. I was thinking for the `value`, since it always seems like a bit of a smell to have a `ReactNode` prop if you're not already using `children` to render content.

[the remark this answers is not in the document]
Comment (jemmyw) [reply src/components/Actions/CardLabel.tsx]: done

**Review by percyhanna — APPROVED** [review]
👍
